bessctl/test_utils.py

```diff
@@ -0,0 +1,243 @@
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import print_function
+
+import os
+import random
+import scapy.all as scapy
+import socket
+import sys
+import unittest
+from time import gmtime, sleep, strftime
+
+try:
+    this_dir = os.path.dirname(os.path.realpath(__file__))
+    sys.path.insert(1, os.path.join(this_dir, '../../../'))
+    from pybess.bess import *
+except ImportError:
+    print('Cannot import the API module (pybess)', file=sys.stderr)
+    raise
+
+SOCKET_PATH = '/tmp/bess_unix_'
+SCRIPT_STARTTIME = strftime("%Y-%m-%d-%H-%M-%S", gmtime())
+
+
+def gen_unix_socket(bess, sockname, timeout_sec=3):
+    socket_port = bess.create_port('UnixSocketPort', sockname,
+                                   {'path': '@' + SOCKET_PATH + sockname})
+    s = socket.socket(socket.AF_UNIX, socket.SOCK_SEQPACKET)
+    s.settimeout(timeout_sec)
+    s.connect('\0' + SOCKET_PATH + sockname)
+
+    return s
+
+
+# generate random packet
+def get_udp_packet(sip=None, dip=None, sport=None, dport=None, pkt_len=60):
+    eth = scapy.Ether(src=scapy.RandMAC()._fix(),
+                      dst=scapy.RandMAC()._fix())
+    ip = scapy.IP(src=sip if sip else scapy.RandIP()._fix(),
+                  dst=dip if dip else scapy.RandIP()._fix())
+    udp = scapy.UDP(sport=sport if sport else random.randrange(pow(2, 16)),
+                    dport=dport if dport else random.randrange(pow(2, 16)))
+    header = eth / ip / udp
+    payload = '0' * (pkt_len - len(header))
+    return header / payload
+
+
+def get_tcp_packet(sip=None, dip=None, sport=None, dport=None, pkt_len=60):
+    eth = scapy.Ether(src=scapy.RandMAC()._fix(),
+                      dst=scapy.RandMAC()._fix())
+    ip = scapy.IP(src=sip if sip else scapy.RandIP()._fix(),
+                  dst=dip if dip else scapy.RandIP()._fix())
+    tcp = scapy.TCP(sport=sport if sport else random.randrange(pow(2, 16)),
+                    dport=dport if dport else random.randrange(pow(2, 16)))
+    header = eth / ip / tcp
+    payload = '0' * (pkt_len - len(header))
+    return header / payload
+
+
+def pkt_str(pkt):
+    if not pkt:
+        return str('').encode('HEX')
+    else:
+        return str(pkt).encode('HEX')
+
+
+class BessModuleAssertion():
+
+    def assertSamePackets(self, pkt1, pkt2):
+        if pkt_str(pkt1) != pkt_str(pkt2):
+            raise AssertionError('"%s" != "%s"',
+                                 pkt_str(pkt1), pkt_str(pkt2))
+
+    def assertNotSamePackets(self, pkt1, pkt2):
+        if pkt_str(pkt1) == pkt_str(pkt2):
+            raise AssertionError('"%s" == "%s"',
+                                 pkt_str(pkt1), pkt_str(pkt2))
+
+    def assertBessAlive(self):
+        self.bess = BESS()
```

Comment:
Recreating (and replacing) a `BESS` instance may be somewhat disruptive. We can either:
1. using existing `self.bess`, we check with a dummy RPC (e.g., `GetVersion`), or
2. we create a new BESS instance, but we do not overwrite self.bess with it.

Reply:
Good point.